### ISSUE TITLE
Validation order: elements go first and collection itself goes second

### ISSUE THREAD

**Issue by abatishchev** [JeremySkinner/FluentValidation#178]
Hi, I have a web api accepting a collection of elements both having validator:

``` csharp
public IHttpActionResult PostData(PaymentCollection payments)
{
}

[Validator(typeof(PaymentCollectionValidator))]
public class PaymentCollection : Collection<Payment>
{
}

[Validator(typeof(PaymentValidator))]
public class Payment
{
}
```

I want to check the collection for length and its elements for validity.

But whrn I make a request, elements go first and collection itself goes second in the validation order even collection is invalid (has too many elements, for instance) so validation is guaranteed to fail.

How can I change the order?

**Comment by JeremySkinner** [comment]
That's not possible if you're using the web api integration. You'd need to invoke the validator manually and use SetValidator instead of using the Validator attributes.

Jeremy

**Comment by abatishchev** [comment]
Can I anyhow extend the existing model validation provider?

**Comment by JeremySkinner** [comment]
Sure - you'd probably want to look at WebApi's DefaultBodyModelValidator as a starting point; this is where the recusrive validation of properties actually occurs.

**Comment by abatishchev** [comment]
So you're suggesting do not try to extend `FluentValidationBodyModelValidator` but `System.Web.Http.Validation.DefaultBodyModelValidator` instead?

**Comment by abatishchev** [comment]
Would you agree to add a public boolean property `ValidateChildrenFirst` to `FluentValidationBodyModelValidator` to control whether or not follow the rule "[Validate the children first - depth-first traversal](https://github.com/JeremySkinner/FluentValidation/blob/master/src/FluentValidation.WebApi/FluentValidationBodyModelValidator.cs#L109)"? e.g.:

```csharp
if (ValidateChildrenFirst)
{
    ValidateChildren();
    isValid = ShallowValidate(metadata, validationContext, container) && isValid;
}
else
{
    isValid = ShallowValidate(metadata, validationContext, container) && isValid;
    ValidateChildren();
}
```

**Comment by JeremySkinner** [comment]
I'm not at all keen on doing that to be honest - the only reason the
fluentvalidationbodymodelvalidator exists is to work around an issue in the
defaultbodymodelvalidator...id prefer to keep its code as close to the
original as possible. Eventually, I'd like to get rid of that class
completely.

If you need this behaviour, I'd suggest having your own Bodymodelvalidator
implementation in your project...so you can then swap out the default (or
the FV one) with your version, if you want to make this modification.

**Comment by abatishchev** [comment]
I see, fair enough. To minimize the pain of copy-pasting the code, could you please change `ValidateNodeAndChildren` to `protected virtual` and other private methods to `protected`?

**Comment by JeremySkinner** [comment]
Honestly I'd rather not for the same reasons...they're private in the
original webapi version and it's literally a copy and paste...I don't want
people relying on the FV version of the class as it's a fork that is going
to go away...I'm sorry if it sounds like I'm being awkward but by opening
it up, I'm then locked into supporting it going forward, which is something
I absolutely don't want to do as its internal webapi code, owned by MS.

Ideally, MS should make that method virtual in the original class (which
would mean that the code wouldn't have had to have been copied in the first
place), but I doubt they'll change it now that aspnet 5 supersedes webapi.

**Comment by abatishchev** [comment]
Can you please elaborate on your reasons do not open it up?
Since Web Api source code [is licensed under the Apache 2.0 license](https://aspnetwebstack.codeplex.com/license) and is released as a separate nuget package, from my point of view, it should be fine. Web Api doesn't go away right away after ASP.NET 6.0 is released, and will be widely used for year after that.

**Comment by JeremySkinner** [comment]
Sure, the license isn't the issue- it's the maintenance and
support overhead.  I understand that from your point of view it's fine,
but by opening it up, it's then something I have to support going forward,
and commit to having it in the codebase for the foreseeable future as
others could be relying on its presence.

Unfortunately I just don't have the capacity to do this...just supporting
FV by itself is already a *massive* time investment for me, and I have to
be very careful about what becomes part of the public api.

Once something is publicly accessible, then I have to commit to supporting
it, because inevitably someone else's code can now be relying on this. It
essentially becomes a contract, in the same way as an interface. Removing
it, or changing it later now becomes a breaking change for all consumers.

This particular class is only in the codebase as a workaround for an issue
with webapi...the aim is that it'll be removed eventually, so I don't want
it to become something that becomes consumed directly or is relied upon. As
this class is essentially a fork, the fewer changes that are made to it the
easier it is for me to maintain. When MS update their version of the class,
the merge process is much easier if I don't have to make lots of small
amendments. If MS completely refactor the class and suddenly change its
structure, I'd have no choice but to make a breaking change to the public
api, and breaking changes Need to be kept to a minimum unless there's a
very good reason.

I know that's not what you want to hear, but this is the reality of
supporting a library that a lot of people rely on, especially when it's a
1-person job

**Comment by abatishchev** [comment]
Sure, I see your point. I can believe that maintaining a heavily used open-source library is a huge deal and chore.
Can I propose two things:
- myself as a co-maintainer of the web api integration package. I'm mainly working with web api projects, for years now and for years to go, I hope.
- since I'm internal to MS, I can contact web api official maintainers and try to convenience them on particular changes. So let's sync on this, what kind of changes you'd like to have in the built-in classes so both your and my goals were achievable.

**Comment by JeremySkinner** [comment]
That would be very helpful, and very much appereciated if you're up for
being involved-thank you.

That'd be especially welcome as I don't personally work with webapi much at
all, and my experience with it is very limited.

I'll document the changes that'd need to happen to support this properly.
Do you mind if I wait until tomorrow morning? I don't have my dev laptop
with me at the moment.

Thanks again

**Comment by abatishchev** [comment]
Awesome, sounds good!
No rush at all, anyway it's the holiday season, so no changes would be reviewed till mid-January, I'm afraid.

**Comment by JeremySkinner** [comment]
So there's a few things that I'd like to see changed in WebApi to better support FV, but for now just focusing on the BodyModelValidator:

If the binder added errors into ModelState before validation occurs, and then FV generates an error message for the same property, the message generated by FV is ignored and never appears. This is because of the following check in DefaultBodyModelValidator.ShallowValidate:

```
if (!validationContext.ModelState.IsValidField(key))
{
  return false;
}
```

The FluentValidationBodyModelValidator changes this check to:
```
var isFluentModelValidator = validator is FluentValidationModelValidator;
...
if ((!isFluentModelValidator && !validationContext.ModelState.IsValidField(key)) ||
	(isFluentModelValidator && validationContext.ModelState.ContainsKey(key))) {
   return false;
}
```
...so it explicitly checks if we're invoking FV, and if so it allows the second error to be added, with a separate key:
```
validationContext.ModelState.AddModelError(isFluentModelValidator ? CreatePropertyModelName(key, error.MemberName) : key, error.Message);
```

Unfortunately, the ShallowValidate method is private static, which is why FluentValidationBodyModelValidator includes so much copy-pasted code. 

The simplest change would be for DefaultBodyModelValidator.ShallowValidate to be virtual, but this would still require some duplicate code. However, this method is nested within several other private static method calls, so could possibly have a large refactoring impact for the webapi team.

**Comment by SeanKilleen** [comment]
Maintenance / cleanup: Closing this as there hasn't been any activity for a few months.

**Comment by abatishchev** [comment]
Actually I work on this :) Just wrote another email yesterday.

**Comment by SeanKilleen** [comment]
@abatishchev is there a reason this issue doesn't link to that work on a branch or PR somewhere? Perhaps we could get them linked together.

From a cleanup perspective, I'm not sure having an issue that was last updated 4 months ago remain open is the best course. Perhaps we could re-phrase the conversation in a new ticket? Or move the conversation from e-mail to this thread so that others can see the progress? 

I'll defer to @JeremySkinner on this one.

**Comment by abatishchev** [comment]
Unfortunately there is no code branch here, yet. However here is a PR in Web Api repo on Codeplex and in parallel I try convince its maintainers to review it. Basically open the validation infrastructure for customization via inheritance, currently it's locked down.

**Comment by JeremySkinner** [comment]
Let's keep this one open for now. I live in hope that the WebApi team will be open to making some changes...

**Comment by SeanKilleen** [comment]
@abatishchev @JeremySkinner makes perfect sense, thanks.

@abatishchev, could you post a link to that PR here so that new and slightly clueless collaborators can follow along? :)

**Comment by abatishchev** [comment]
Oh, yes, absolutely. https://aspnetwebstack.codeplex.com/SourceControl/network/forks/abatishchev/modelvalidation2/contribution/8644
Thanks for support!

**Comment by JeremySkinner** [comment]
I just saw that there was actually a commit made to the WebApi code to enable this in October (only 2 years late!) Not sure how relevant this is now that that we have aspnetcore, but it looks like this will be part of WebApi 5.2.4 so we could look at doing this once WebApi 5.2.4 is out. @abatishchev Would you be up for doing a pull request for this once the aspnet team finish their release?

**Comment by abatishchev** [comment]
Wow, really? I literally lost hope. Let me check that out, will come back shortly.

**Comment by abatishchev** [comment]
Didn't know they resumed the work on WebApi, and for quite some time. Sure, I'll monitor and will pick the changes once they're out.

**Comment by abatishchev** [comment]
These changes are already available from nightly build https://dotnet.myget.org/feed/aspnetwebstack-dev/package/nuget/Microsoft.AspNet.WebApi so I'll give it a try

**Comment by abatishchev** [comment]
@JeremySkinner FluentValidation.WebApi in master depends on AspNet.Core. Shouldn't it be FluentValidation.WebApi which still depends on AspNet.WebApi and FluentValidation.WebApi.Core which would depend on AspNet.Core?

**Comment by abatishchev** [comment]
But maybe I'm wrong and just out of date with modern projects structure.

**Comment by JeremySkinner** [comment]
It doesn’t depend on aspnetcore - it’s just using the new SDK-based projects. Still targets .net 45 and webapi2 as always, but used the modern build chain.

**Comment by abatishchev** [comment]
Right, I got confused with Core in package name, forgot that's how WebApi is structured. Please disregard.

**Comment by abatishchev** [comment]
By doing `FluentValidationBodyModelValidator : DefaultBodyModelValidator` with no overrides I have all tests but 1 succeeded. Looking into it: `Should_still_validate_other_properties_when_error_found_in_collection`

**Comment by abatishchev** [comment]
The test fails because the built-in provider doesn't bother to validate the model if its elements aren't valid, see https://github.com/aspnet/AspNetWebStack/blob/master/src/System.Web.Http/Validation/DefaultBodyModelValidator.cs#L162.

So our options are:

- Ask @dougbu to allow the flexibility in the base class
- Override just `ValidateNodeAndChildren()` what is now possible. Yay! However it's huge and uses internal helper. Implement it differently, allow the flexibility by our own.

**Comment by JeremySkinner** [comment]
Yes that's the same behaviour as aspnetcore (not validating the root model if any children fail). The aspnet team recently merged my pull request in to allow this behaviour to be customized (see github.com/aspnet/mvc/issues/7014) so I expect they'd be open to doing the same in webapi.

Although saying that....I think we can achieve it pretty easily by just overriding the method. I don't think we should need to implement its internals. Pseudocode example:

```
override ValidateNodeAndChildren(metadata, validationContext, container, validators) {
   bool isValid = base.ValidateNodeAndChildren(metadata, validationContext, container, validators); 

   if(! isValid && !HasAlreadyBeenValidated(container)) {
       var rootIsValid = ShallowValidate(metadata, validationContext, container, validators);
       return rootIsValid && isValid;
   }
   return isValid; 
}
```
....where `HasAlreadyBeenValidated` should be able to check the root model has been validated or not (either by checking ModelState, or if that's not possible then the FluentValidationModelValidatorProvider could track the objects it validates in an HttpContext.Items variable which we could check).

**Comment by abatishchev** [comment]
Sorry for a pause in response, it's a holidays season, you know, one day no work, another day a ton.

The issue I see in calling `base.ValidateNodeAndChildren()` that it'll validate children first, What might be relatively resource-intensive. And only then will validate the root model. Ideally we shoudn't validate them at all rather than do this but swallow the errors (if any). I'd keep this for plan B if there will be no better options.

**Comment by abatishchev** [comment]
@JeremySkinne how do you think which option we should perused first? open up the build-in provider for maximum flexibility/extensibility/customization?

**Comment by JeremySkinner** [comment]
I will try and do some prototyping and will get back to you

**Comment by abatishchev** [comment]
@JeremySkinner did you have a chance to take a look and try out few options here?

**Comment by JeremySkinner** [comment]
Not yet sorry. I haven’t had much tune to work on FluentValidation and most of it has been spent fixing aspnetcore issues.

**Comment by JeremySkinner** [comment]
I took a bit of time today to look at this and made the following changes:
- Updated FluentValidation.WebApi to depend on WebApi 5.2.4-preview1 
- Changed FluentValidationBodyModelValidator to inherit from 
- Implemented the change I mentioned above with overriding ValidateNodeAndChildren

This is checked in. 

The potential performance issue with validating children first is not a concern - this is the standard behaviour in all the .net model binding/validation routines (MVC5, webapi and aspnetcore) so we should keep this as the default behaviour for consistency with both standard validation as well as FV's own integrations with MVC5 and AspNetCore (which all follow the same pattern). 

In your case where you explicitly want to validate the root before its children then you'd do this by overriding `ValidateNodeAndChildren` and explicitly move the call to ShallowValidate on the root model to happen before the calls to ValidateProperties/ValidateElements.

**Comment by abatishchev** [comment]
Awesome! I think it's good enough. I'll watch closely for updates of the WebApi package and will keep you posted here, or elsewhere. Thanks again for your work and help on this issues.

**Comment by JeremySkinner** [comment]
WebApi 5.2.4 has been out for a few weeks now, so I've done a new release of the FluentValidation.WebApi package (7.4.0) that contains the updated FluentValidationBodyModelValidator and the other changes that we discussed.

Let me know if you run into any issues.
